### PR TITLE
[NTOSKRNL] Local Inter-Process Communication (LPC) improvements

### PR DESCRIPTION
**Pull request by belliash**
## Purpose

Several changes, fixes and improvements to the Local Inter-Process Communication (LPC) facility provided by the kernel. This as well includes some missing parts of LPC implementation and makes it more complete.

JIRA issue: [CORE-17075](https://jira.reactos.org/browse/CORE-17075)

## Proposed changes
- Add more debugging messages,
- Add more sanity checks (eg. to make sure reply message is on the reply chain),
- Implement and make use of LpcWaitablePortObjectType,
- Implement NtReplyWaitReplyPort(),
- Implement NtQueryInformationPort(),
- Implement callbacks support,
- Implement Thread Semaphore support,
- Unify the way of checking message type by using LpcpGetMessageType(),
- Several small fixes and improvements

### CODE REVIEW

File: ntoskrnl/lpc/port.c [hunk 1]
```diff
@@ -277,7 +276,7 @@ NTAPI
 NtQueryPortInformationProcess(VOID)
 {
     /* This is all this function does */
-    return STATUS_UNSUCCESSFUL;
+    return TRUE;
```
Comment (Fraizeraust) [inline ntoskrnl/lpc/port.c:279]: If this function really returns a boolean value then you must change the function's declaration to a `BOOLEAN` as well.

Comment (belliash) [reply ntoskrnl/lpc/port.c:279]: Yes, you are right. And this should be fixed in NDK as well.

Comment (HBelusca) [reply ntoskrnl/lpc/port.c:279]: All documentation (including public one like MSDN) mentions that this function's return type is NTSTATUS.

File: ntoskrnl/lpc/port.c [hunk 2]
```diff
@@ -287,8 +287,40 @@ NtQueryInformationPort(IN HANDLE PortHandle,
                        IN ULONG PortInformationLength,
                        OUT PULONG ReturnLength)
 {
-    UNIMPLEMENTED;
-    return STATUS_NOT_IMPLEMENTED;
+	KPROCESSOR_MODE PreviousMode;
```
Comment (HBelusca) [inline ntoskrnl/lpc/port.c:290]: Please fix the whole code style at once (including the 4-space indentation) in this commit (+squash together with the next commit).

File: ntoskrnl/lpc/reply.c [hunk 3]
```diff
@@ -864,8 +864,8 @@ NtReplyWaitReplyPort(IN HANDLE PortHandle,
     PAGED_CODE();
     LPCTRACE(LPC_REPLY_DEBUG,
              "%s Attempted reply wait reply (Handle: %p, Message: %p)\n",
-	         PsGetCurrentProcess()->ImageFileName,
-	         PortHandle,
+             PsGetCurrentProcess()->ImageFileName,
+             PortHandle,
```
Comment (HBelusca) [inline ntoskrnl/lpc/reply.c:868]: Maybe squash with the previous commit? (the one modifying some other LPCTRACEs)

File: ntoskrnl/lpc/port.c (hunk 1 above)
Comment (belliash) [reply ntoskrnl/lpc/port.c:279]: Yes, I have found out that this function returns "1", thus TRUE seemed the valid return value here. I have not verified the function declaration to be honest, but changed the type as suggested by @Fraizeraust. This seems to be my mistake, as I googled for that now and indeed seems that it should be NTSTATUS...

File: ntoskrnl/lpc/reply.c (hunk 3 above)
Comment (belliash) [reply ntoskrnl/lpc/reply.c:868]: Should look better now.

File: ntoskrnl/lpc/complete.c [hunk 4]
```diff
@@ -275,7 +278,8 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
     if (ClientSectionToMap)
     {
         /* Setup the offset */
-        SectionOffset.QuadPart = ConnectMessage->ClientView.SectionOffset;
+        SectionOffset.LowPart = ConnectMessage->ClientView.SectionOffset;
```
Comment (Extravert-ir) [inline ntoskrnl/lpc/complete.c:281]: Why this change?

**Comment by Extravert-ir** [comment]
Do we have any kind of tests for LPC? Would be nice to check what started to work now

File: ntoskrnl/lpc/complete.c [hunk 5]
```diff
@@ -315,11 +319,85 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
...
 
     /* Check if there's a server section */
-    if (ServerView)
+    if (NT_SUCCESS(Status) && ServerView)
     {
-        /* FIXME: TODO */
-        UNREFERENCED_PARAMETER(CapturedServerView);
-        ASSERT(FALSE);
+        SectionOffset.LowPart = CapturedServerView.SectionOffset;
+        SectionOffset.HighPart = 0;
```
Comment (HBelusca) [inline ntoskrnl/lpc/complete.c:325]: See question from above.

File: ntoskrnl/lpc/complete.c [hunk 6]
```diff
@@ -315,11 +319,85 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
...
+                if (ServerPort->MappingProcess == NULL)
+                {
+                    ServerPort->MappingProcess = PsGetCurrentProcess();
+                    ObReferenceObject(ServerPort->MappingProcess);
+                }
+                CapturedServerView.SectionOffset = SectionOffset.LowPart;
+                CapturedServerView.ViewBase = ServerPort->ServerSectionBase;
+
+                SectionOffset.LowPart = CapturedServerView.SectionOffset;
+                SectionOffset.HighPart = 0;
```
Comment (HBelusca) [inline ntoskrnl/lpc/complete.c:361]: Ditto

File: ntoskrnl/lpc/complete.c [hunk 7]
```diff
@@ -315,11 +319,85 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
...
+
+            if (NT_SUCCESS(Status))
+            {
+                /* Section mapped into server process */
+                if (ServerPort->MappingProcess == NULL)
+                {
+                    ServerPort->MappingProcess = PsGetCurrentProcess();
+                    ObReferenceObject(ServerPort->MappingProcess);
+                }
+                CapturedServerView.SectionOffset = SectionOffset.LowPart;
```
Comment (HBelusca) [inline ntoskrnl/lpc/complete.c:357]: Ditto.

File: ntoskrnl/lpc/complete.c [hunk 8]
```diff
@@ -469,6 +547,25 @@ NtCompleteConnectPort(IN HANDLE PortHandle)
         return STATUS_SUCCESS;
     }
 
+    /* Ensure client thread is on the reply chain */
+    if (Port->ConnectionPort)
+    {
+        PLIST_ENTRY Entry;
+        for (Entry = Port->ConnectionPort->LpcReplyChainHead.Flink; Entry != (PLIST_ENTRY)(&Port->ConnectionPort->LpcReplyChainHead.Flink); Entry = Entry->Flink)
```
Comment (HBelusca) [inline ntoskrnl/lpc/complete.c:554]: This line is so long, that you can basically place each part of the for-loop on its own line.

File: ntoskrnl/lpc/connect.c [hunk 9]
```diff
@@ -392,7 +406,8 @@ NtSecureConnectPort(OUT PHANDLE PortHandle,
         }
 
         /* Set the section offset */
-        SectionOffset.QuadPart = CapturedClientView.SectionOffset;
+        SectionOffset.LowPart = CapturedClientView.SectionOffset;
+        SectionOffset.HighPart = 0;
```
Comment (HBelusca) [inline ntoskrnl/lpc/connect.c:410]: See question from above.

File: ntoskrnl/lpc/port.c [hunk 10]
```diff
@@ -288,8 +287,45 @@ NtQueryInformationPort(IN HANDLE PortHandle,
...
+    PLPCP_PORT_OBJECT PortObject;
+
+    PAGED_CODE();
+
+    UNREFERENCED_PARAMETER(PortInformationClass);
+
+    PreviousMode = KeGetPreviousMode();
+    if (PreviousMode != KernelMode)
+    {
+        _SEH2_TRY {
```
Comment (HBelusca) [inline ntoskrnl/lpc/port.c:301]: Please fix the code style of these SEH blocks.

File: ntoskrnl/lpc/port.c [hunk 11]
```diff
@@ -288,8 +287,45 @@ NtQueryInformationPort(IN HANDLE PortHandle,
...
+        {
+            Status = ObReferenceObjectByHandle(PortHandle, GENERIC_READ, LpcWaitablePortObjectType, PreviousMode, (PVOID*)&PortObject, NULL);
+            if (!NT_SUCCESS(Status))
+            {
+                return Status;
+            }
+        }
+        ObDereferenceObject(PortObject);
+        return STATUS_SUCCESS;
+    } else {
```
Comment (HBelusca) [inline ntoskrnl/lpc/port.c:326]: Code style.

File: ntoskrnl/lpc/send.c [hunk 12]
```diff
@@ -228,7 +230,13 @@ LpcRequestWaitReplyPort(IN PVOID PortObject,
             return STATUS_INVALID_PARAMETER;
     }
 
-    /* Set the request type */
+    /* Pass LPC_KERNELMODE_MESSAGE bit in message type field */
+    if (LpcRequest->u2.s2.Type & LPC_KERNELMODE_MESSAGE)
+    {
+        MessageType |= LPC_KERNELMODE_MESSAGE;
+	}
```
Comment (HBelusca) [inline ntoskrnl/lpc/send.c:237]: Broken indentation.

File: ntoskrnl/lpc/send.c [hunk 13]
```diff
@@ -228,7 +230,13 @@ LpcRequestWaitReplyPort(IN PVOID PortObject,
...
+
+	/* Set the request type */
```
Comment (HBelusca) [inline ntoskrnl/lpc/send.c:239]: Idem?

File: ntoskrnl/lpc/send.c [hunk 14]
```diff
@@ -250,10 +258,105 @@ LpcRequestWaitReplyPort(IN PVOID PortObject,
...
+        /* Allocate and initialize request message */
+        LpcpMoveMessage(&Message->Request,
+                        LpcRequest,
+                        LpcRequest + 1,
+                        0,
+                        &Thread->Cid);
+
+        /* Generate the Callback ID and set it */
+        Message->Request.CallbackId = LpcpNextCallbackId++;
+        if(!LpcpNextCallbackId) LpcpNextCallbackId = 1;
```
Comment (HBelusca) [inline ntoskrnl/lpc/send.c:328]: space after "if".

File: ntoskrnl/lpc/send.c [hunk 15]
```diff
@@ -830,9 +976,128 @@ NtRequestWaitReplyPort(IN HANDLE PortHandle,
...
+            /* Check if this is a communication port */
+            if ((Port->Flags & LPCP_PORT_TYPE_MASK) == LPCP_CLIENT_PORT)
+            {
+                Message->PortContext = ReplyPort->PortContext;
+            }
+        }
+
+        /* Generate the Callback ID and set it */
+        Message->Request.CallbackId = LpcpNextCallbackId++;
+        if(!LpcpNextCallbackId) LpcpNextCallbackId = 1;
```
Comment (HBelusca) [inline ntoskrnl/lpc/send.c:1069]: space after "if".

File: ntoskrnl/lpc/reply.c [hunk 16]
```diff
@@ -754,15 +847,278 @@ NtReplyWaitReceivePort(IN HANDLE PortHandle,
...
+
+    PAGED_CODE();
+    LPCTRACE(LPC_REPLY_DEBUG,
+             "%s Attempted reply wait reply (Handle: %p, Message: %p)\n",
+             PsGetCurrentProcess()->ImageFileName,
+             PortHandle,
+             ReplyMessage);
+
+    /* Check if the call comes from user mode */
+    if (PreviousMode != KernelMode) {
```
Comment (HBelusca) [inline ntoskrnl/lpc/reply.c:872]: Code style.

File: ntoskrnl/lpc/reply.c [hunk 17]
```diff
@@ -754,15 +847,278 @@ NtReplyWaitReceivePort(IN HANDLE PortHandle,
...
+            {
+                LpcpMoveMessage(ReplyMessage,
+                                &Message->Request,
+                                (&Message->Request) + 1,
+                                0,
+                                NULL);
+            }
+            _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+            {
+                DPRINT1("Got exception!\n");
```
Comment (HBelusca) [inline ntoskrnl/lpc/reply.c:1088]: Useless unless you also print the exception code.

File: ntoskrnl/lpc/send.c [hunk 18]
```diff
@@ -830,9 +976,128 @@ NtRequestWaitReplyPort(IN HANDLE PortHandle,
...
+                DPRINT1("LpcpVerifyMessageDataInfo failed\n");
+                LpcpFreeToPortZone(Message, LPCP_LOCK_HELD | LPCP_LOCK_RELEASE);
+                ObDereferenceObject(WakeupThread);
+                ObDereferenceObject(Port);
+                _SEH2_YIELD(return Status);
+            }
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            DPRINT1("Got exception\n");
```
Comment (HBelusca) [inline ntoskrnl/lpc/send.c:1031]: Useless unless you also print the exception code.

File: ntoskrnl/lpc/complete.c (hunk 4 above)
Comment (HBelusca) [reply ntoskrnl/lpc/complete.c:281]: The question I have, when I see this (and the other instances) is whether these changes break x64 compatibility (indeed in x64 an offset could theoretically be larger than what can be held in a DWORD.)

Comment (belliash) [reply ntoskrnl/lpc/complete.c:281]: This is just small compiler compatibility improvement. If the compiler does not have built-in support for 64-bit integers, you should always use the LowPart and HighPart quantities instead of QuadPart.

File: ntoskrnl/lpc/complete.c (hunk 5 above)
Comment (belliash) [reply ntoskrnl/lpc/complete.c:325]: See answer above.

File: ntoskrnl/lpc/complete.c (hunk 7 above)
Comment (belliash) [reply ntoskrnl/lpc/complete.c:357]: See answer above.

File: ntoskrnl/lpc/complete.c (hunk 6 above)
Comment (belliash) [reply ntoskrnl/lpc/complete.c:361]: See answer above.

File: ntoskrnl/lpc/connect.c (hunk 9 above)
Comment (belliash) [reply ntoskrnl/lpc/connect.c:410]: See answer above.

File: ntoskrnl/lpc/complete.c (hunk 8 above)
Comment (belliash) [reply ntoskrnl/lpc/complete.c]: Corrected.

File: ntoskrnl/lpc/reply.c (hunk 16 above)
Comment (belliash) [reply ntoskrnl/lpc/reply.c]: Corrected.

File: ntoskrnl/lpc/send.c (hunk 15 above)
Comment (belliash) [reply ntoskrnl/lpc/send.c]: Corrected.

File: ntoskrnl/lpc/complete.c (hunk 4 above)
Comment (binarymaster) [reply ntoskrnl/lpc/complete.c:281]: Do we use any compilers that don't support QuadPart? If not, this is not needed, unless you have some plans on it (then please explain them).

File: ntoskrnl/lpc/send.c (hunk 18 above)
Comment (belliash) [reply ntoskrnl/lpc/send.c:1030]: Right.

File: ntoskrnl/lpc/reply.c (hunk 17 above)
Comment (belliash) [reply ntoskrnl/lpc/reply.c]: Right.

File: ntoskrnl/lpc/complete.c (hunk 4 above)
Comment (Extravert-ir) [reply ntoskrnl/lpc/complete.c:281]: We don't support compilers without `long long` support, so please use such pattern only when needed by logic

Comment (HBelusca) [reply ntoskrnl/lpc/complete.c:281]: A ton of code in our code base requires this long long support be present already (and in other parts of the kernel), so there is no need to keep that.

Comment (belliash) [reply ntoskrnl/lpc/complete.c:281]: Microsoft defines LARGE_INTEGER in Windows NT as data type that is 64-bit integer created from an array of two longwords. It is not the same as quadword and this is how I used it.

Comment (belliash) [reply ntoskrnl/lpc/complete.c:281]: This is still true for at least NT 5.2. Maybe newer versions changes that, dunno. Anyway, will apply your guidelines in free time.

File: ntoskrnl/lpc/complete.c [hunk 19]
```diff
@@ -469,6 +547,27 @@ NtCompleteConnectPort(IN HANDLE PortHandle)
         return STATUS_SUCCESS;
     }
 
+    /* Ensure client thread is on the reply chain */
+    if (Port->ConnectionPort)
+    {
+        PLIST_ENTRY Entry;
+        for (Entry = Port->ConnectionPort->LpcReplyChainHead.Flink;
+             Entry != (PLIST_ENTRY)(&Port->ConnectionPort->LpcReplyChainHead.Flink);
```
[the remark this answers is not in the document]
Comment (belliash) [reply ntoskrnl/lpc/complete.c:555]: Casting here was added to fix compiler error about comparison of distinct pointer types.

[the remark this answers is not in the document]
Comment (HBelusca) [reply ntoskrnl/lpc/complete.c:555]: But there are not distinct, since LpcReplyChainHead is LIST_ENTRY and thus, &LpcReplyChainHead is a PLIST_ENTRY, and LpcReplyChainHead.Flink is already a PLIST_ENTRY.

File: ntoskrnl/include/internal/lpc_x.h [hunk 20]
```diff
@@ -115,6 +115,7 @@ LpcpAllocateFromPortZone(VOID)
     InitializeListHead(&Message->Entry);
     Message->RepliedToThread = NULL;
     Message->Request.u2.ZeroInit = 0;
+    Message->Request.u2.s2.Type = 0;
```
[the remark this answers is not in the document]
Comment (belliash) [reply ntoskrnl/include/internal/lpc_x.h]: Indeed this is union.

File: ntoskrnl/ntoskrnl.spec [hunk 21]
```diff
@@ -740,6 +740,7 @@
 @ stdcall LdrFindResourceDirectory_U(ptr ptr long ptr)
 @ stdcall LdrFindResource_U(ptr ptr long ptr)
 @ extern LpcPortObjectType
+@ extern LpcWaitablePortObjectType
```
[the remark this answers is not in the document]
Comment (belliash) [reply ntoskrnl/ntoskrnl.spec]: Seems you are right!

File: ntoskrnl/lpc/complete.c (hunk 4 above)
Comment (belliash) [reply ntoskrnl/lpc/complete.c]: Fixed.

File: ntoskrnl/lpc/port.c [hunk 22]
```diff
@@ -288,8 +287,49 @@ NtQueryInformationPort(IN HANDLE PortHandle,
...
+    {
+        _SEH2_TRY
+        {
+            ProbeForWrite(PortInformation, PortInformationLength, sizeof(ULONG));
+            if (ARGUMENT_PRESENT(ReturnLength))
+            {
+                ProbeForWriteUlong(ReturnLength);
+            }
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER) {
```
Comment (HBelusca) [inline ntoskrnl/lpc/port.c:309]: Code style.

File: ntoskrnl/lpc/port.c [hunk 23]
```diff
@@ -288,8 +287,49 @@ NtQueryInformationPort(IN HANDLE PortHandle,
...
+            _SEH2_YIELD(return _SEH2_GetExceptionCode());
+        }
+        _SEH2_END;
+    }
+
+    if (ARGUMENT_PRESENT(PortHandle))
+    {
+        Status = ObReferenceObjectByHandle(PortHandle, GENERIC_READ, LpcPortObjectType, PreviousMode, (PVOID*)&PortObject, NULL);
```
Comment (HBelusca) [inline ntoskrnl/lpc/port.c:317]: Kernel code style likes to have each argument of functions that take a ton of arguments (like this one), on their own lines.

File: ntoskrnl/lpc/port.c [hunk 24]
```diff
@@ -288,8 +287,49 @@ NtQueryInformationPort(IN HANDLE PortHandle,
...
+
+    UNREFERENCED_PARAMETER(PortInformationClass);
+
+    PreviousMode = KeGetPreviousMode();
+    if (PreviousMode != KernelMode)
+    {
+        _SEH2_TRY
+        {
+            ProbeForWrite(PortInformation, PortInformationLength, sizeof(ULONG));
+            if (ARGUMENT_PRESENT(ReturnLength))
```
Comment (HBelusca) [inline ntoskrnl/lpc/port.c:304]: We don't use this macro in the kernel.

File: ntoskrnl/lpc/complete.c [hunk 25]
```diff
@@ -315,11 +318,83 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
...
+                }
+            }
+            else
+            {
+                ObDereferenceObject(ServerPort);
+            }
+            ObDereferenceObject(SectionToMap);
+        }
+        else
+        {
```
Comment (HBelusca) [inline ntoskrnl/lpc/complete.c:395]: I think that all these repetitive `ObDereferenceObject(ServerPort);` could be written just once, with a separate block:
```c
if (!NT_SUCCESS(Status))
    ObDereferenceObject(ServerPort);
```
at the end of the whole code of this `if (NT_SUCCESS(Status) && ServerView)` big block.
(Since all the code inside it is a nesting of "do_a_call; if (NT_SUCCESS(Status)) {do_something_else;}".)

File: ntoskrnl/lpc/port.c (hunk 24 above)
Comment (ThFabba) [reply ntoskrnl/lpc/port.c:304]: Well maybe we should...

File: ntoskrnl/lpc/complete.c [hunk 26]
```diff
@@ -160,7 +162,8 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
     /* Make sure that the client wants a reply, and this is the right one */
     if (!(LpcpGetMessageFromThread(ClientThread)) ||
         !(CapturedReplyMessage.MessageId) ||
-        (ClientThread->LpcReplyMessageId != CapturedReplyMessage.MessageId))
+        (ClientThread->LpcReplyMessageId != CapturedReplyMessage.MessageId) ||
+        (LpcpGetMessageType(&LpcpGetMessageFromThread(ClientThread)->Request) != LPC_CONNECTION_REQUEST))
```
[the remark this answers is not in the document]
Comment (belliash) [reply ntoskrnl/lpc/complete.c:169]: Yes, no need to call LpcpGetMessageFromThread() twice.

File: ntoskrnl/lpc/close.c [hunk 27]
```diff
@@ -346,6 +356,7 @@ LpcpDeletePort(IN PVOID ObjectBody)
         ClientDiedMsg.h.u1.s1.TotalLength = sizeof(ClientDiedMsg);
         ClientDiedMsg.h.u1.s1.DataLength = sizeof(ClientDiedMsg.CreateTime);
         ClientDiedMsg.h.u2.ZeroInit = 0;
+        ClientDiedMsg.h.u2.s2.DataInfoOffset = 0;
```
[the remark this answers is not in the document]
Comment (belliash) [reply ntoskrnl/lpc/close.c]: I think I already saw this...

File: ntoskrnl/lpc/port.c (hunk 23 above)
Comment (belliash) [reply ntoskrnl/lpc/port.c]: Fix.

File: ntoskrnl/lpc/port.c (hunk 22 above)
Comment (belliash) [reply ntoskrnl/lpc/port.c]: Fix.

File: ntoskrnl/lpc/connect.c [hunk 28]
```diff
@@ -733,6 +747,15 @@ NtSecureConnectPort(OUT PHANDLE PortHandle,
     return Status;
 
 Failure:
+    if (KeReadStateSemaphore(&Thread->LpcReplySemaphore))
```
Comment (ThFabba) [inline ntoskrnl/lpc/connect.c:750]: As far as I can tell:
* The STATUS_PORT_CONNECTION_REFUSED case above (L726/740) should not wait on the semaphore
* The first failure case in L598/612 already does this wait
* The exception case in L631/645 should not wait (actually to me it looks like it shouldn't even `goto Failure`)

So this doesn't seem right. Can you double-check?

File: ntoskrnl/lpc/complete.c [hunk 29]
```diff
@@ -159,8 +159,11 @@ NtAcceptConnectPort(OUT PHANDLE PortHandle,
...
 
+    /* Get the message */
+    Message = LpcpGetMessageFromThread(ClientThread);
+
     /* Make sure that the client wants a reply, and this is the right one */
-    if (!(LpcpGetMessageFromThread(ClientThread)) ||
+    if (!(Message) ||
         !(CapturedReplyMessage.MessageId) ||
         (ClientThread->LpcReplyMessageId != CapturedReplyMessage.MessageId) ||
         (LpcpGetMessageType(&LpcpGetMessageFromThread(ClientThread)->Request) != LPC_CONNECTION_REQUEST))
```
Comment (HBelusca) [inline ntoskrnl/lpc/complete.c:169]: There is still an instance of this call here in the test.

File: ntoskrnl/lpc/reply.c [hunk 30]
```diff
@@ -45,16 +45,30 @@ LpcpFreeDataInfoMessage(IN PLPCP_PORT_OBJECT Port,
             (Message->Request.ClientId.UniqueThread == ClientId.UniqueThread) &&
             (Message->Request.ClientId.UniqueProcess == ClientId.UniqueProcess))
         {
+            LPCTRACE(LPC_REPLY_DEBUG,
+                     "%s removing DataInfo Message %lx (%u.%u) Port: $lx\n",
```
Comment (ThFabba) [inline ntoskrnl/lpc/reply.c:49]: Typo: `$` instead of `%`.

Comment (ThFabba) [inline ntoskrnl/lpc/reply.c:49]: `Message %p` please since `Message` is a pointer.

File: ntoskrnl/lpc/reply.c [hunk 31]
```diff
@@ -83,6 +97,14 @@ LpcpSaveDataInfoMessage(IN PLPCP_PORT_OBJECT Port,
         }
     }
 
+    LPCTRACE(LPC_REPLY_DEBUG,
+             "%s saving DataInfo Message %lx (%u.%u) Port: %lx\n",
```
Comment (ThFabba) [inline ntoskrnl/lpc/reply.c:101]: %p please

File: ntoskrnl/lpc/complete.c (hunk 25 above)
Comment (belliash) [reply ntoskrnl/lpc/complete.c:397]: Yes, probably. Feel free to refactor this.

**Comment by ColinFinck** [comment]
We got evidence that your Pull Request closely resembles the copyrighted code of the Windows Research Kernel beyond reasonable doubt.
Among other things, the [WRK Source Code License](https://web.archive.org/web/20081202142504/http://www.microsoft.com:80/resources/sharedsource/licensing/basics/wrklicense.mspx) explicitly restricts usage of the Windows Research Kernel to "educational institution[s]" and only permits redistribution "to other eligible users of the software".

We are aware of the existence of the Windows Research Kernel and have therefore explicitly mentioned it in our [Contribution Guidelines](https://github.com/reactos/reactos/blob/master/CONTRIBUTING.md):

> Before contributing, you must affirm that the following is true:
> I hereby swear that I have not used nor seen the source code to any version of the Windows operating system nor any Microsoft product that may be related to the proposed project that is under a license incompatible with contribution to ReactOS, including but not limited to the leaked Windows 2000 source code and the Windows Research Kernel.

By trying to sneak in your PR into the ReactOS Project and thereby violating the WRK license and our contribution guidelines, you have put the project at legal risk.
We will therefore close this PR and ban you from any future contributions to ReactOS.

**Comment by veselcraft** [comment]
bruh
